### PR TITLE
Unify Array APIs

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes:https://github.com/composewell/streamly/issues/2334

File: core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs [hunk 1]
```diff
@@ -27,6 +27,11 @@ module Streamly.Internal.Data.Array.Generic.Mut.Type
     , writeN
     , writeWith
     , write
+    , fromStreamN
+    , fromStream
+
+    , fromListN
+    , fromList
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs:34]: What is the reason for exposing these? I do need see these used or exported from any exposed module. Are these supposed to be exposed from the released module?

File: core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs [hunk 2]
```diff
@@ -529,12 +535,12 @@ toList arr@MutArray{..} = mapM (`getIndexUnsafe` arr) [0 .. (arrLen - 1)]
...
 --
--- @toStreamD = D.unfold read@
+-- @read = D.unfold read@
 --
 -- We can try this if the unfold has any performance issues.
-{-# INLINE_NORMAL toStreamD #-}
-toStreamD :: MonadIO m => MutArray a -> D.Stream m a
-toStreamD arr@MutArray{..} =
+{-# INLINE_NORMAL read #-}
+read :: MonadIO m => MutArray a -> D.Stream m a
```
[the remark this answers is not in the document]
Comment (rnjtranjan) [reply core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs:535]: it uses "getIndexUnsafe" which requires a MonadIO constraint.

File: core/src/Streamly/Data/MutArray/Generic.hs [hunk 3]
```diff
@@ -35,15 +35,22 @@ module Streamly.Data.MutArray.Generic
...
     , putIndex
     , modifyIndex
+
+    -- * Size
+    , length
     )
 where
 
 import Streamly.Internal.Data.Array.Generic.Mut.Type
+import  Prelude hiding(read, length)
```
Comment (adithyaov) [inline core/src/Streamly/Data/MutArray/Generic.hs:54]:
```suggestion
import  Prelude hiding (read, length)
```

File: core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (adithyaov) [reply core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs:536]: This is a mutable array. This needs to have a MonadIO constraint.

File: core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs [hunk 4]
```diff
@@ -638,6 +638,26 @@ writeWith elemCount = FL.rmapM extract $ FL.foldlM' step initial
...
+-- array is allocated to size @n@, if the stream terminates before @n@
+-- elements then the array may hold less than @n@ elements.
+--
+{-# INLINE fromStreamN #-}
+fromStreamN :: MonadIO m => Int -> Stream m a -> m (MutArray a)
+fromStreamN n = D.fold (writeN n)
+
+{-# INLINE fromStream #-}
+fromStream :: MonadIO m => Stream m a -> m (MutArray a)
+fromStream = D.fold write
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Array/Generic/Mut/Type.hs:651]: Do we need these? INLINE these definations?

**Pull request merged**